content/knowledge_base/flare/consensus/avalanche/specification.md

```diff
@@ -0,0 +1,210 @@
+Additionally each node maintains its own local list of known nodes: Nu — nodes known to `u` is a subset of the set of all existing nodes: `Nu ⊆ N`.
+
+Each node implements an event-driven state machine, centered around a query that serves both to solicit votes on each transaction and to notify other nodes of the existence of newly discovered transactions.
+When a node `u` discovers a trancation `T` through a query, it starts a one-time query process by sampling k random peers and sending a message to them, after T is delivered to the node by onReceiveTx.
+Node `u` answers a query by checking whether each T' such that `T' *<— T` is currently prefered among competing transactions `∀T'' ∈ PT'` from conflict set P of T'.
+If every single ancestor T' fulfills this criterion, the transation is said to br strongly preferred and receives a yes-vote (1).
+A failure on this criterion at any T' yields a no-vote (0).
+When `u` accumulates k responses, it checks whether there are α yes-votes for T, and if so grants the chit value (cT = 1).
+This process will yield a labeling of the DAG with a chit value and associated confidence for each transaction T.
+
+Example of (chit,confidence) values:
+
+```text
+
+T1(1,6)
+├── T2(1,5)
+│    ├── T4(1,2) ── T8(1,1)
+│    │               │
+│    └────────────── T5(1,3)
+│                    └──── T9(1,1)
+│
+└── T3(0,0)
+    ├── T6(0,0)
+    └── T7(0,0)
+
+```
+
+Conflict set P for T1 has one element T1.
+Conflict set P for T2 and T3 has two elements {T2, T3}
+Conflict set P for T5, T6 and T9 has 3 elements {T6, T7, T9}
+
+Similar to Snowball, sampling in Avalanche will create a positive feedback for the preference of a single transaction in its conflict set.
+When a transaction has a larger confidence its decendants are likely to collect chits in the future compared to others with lower confidence.
+For example T2 has larger confidence than T3 and its decendants are more likely to collect chits in the future.
+
+Similar to Bitcoin, Avalanche leaves determining the acceptance point of a transaction to the application.
+Committing a transaction can be performed through a safe early commitment.
+For virtuous transactions, T is accepted when it is the only transaction in its confliict set and has a confidence no less than a threshold of β1.
+If a virtuous transaction fails to get accepted due to a problem with parents, it could be accepted if reissued with different parents.
+
+### Transaction Entangling
+
+How Avalance entangles transactions:
+
+```text
+init:
+    `Set T` = Ø set of known transactions (empty set)
+    Q = Ø set of queried transactions (empty set)
+
+generateTx:
+    edges = {T' <— T: T' ∈ parentSelection(`Set T`)}
+    T = Tx(data, edges)
+    onReceiveEvent(T)
+
+onReceiveEvent(t)
+    if T ∉ `Set T` then
+        if PT (conflict set of T) = Ø then
+            PT = {T}, Pt.pref = T
+            PT.last = T, PT.cnt = 0
+        else PT = PT U {T}
+        `Set T` = T U {T}, cT = 0
+```
```

Comment:
What is this syntax? Is it pseudocode? It kind of looks like Pascal-style pseudocode so I don't really know which syntax highlighter to suggest using here.

Reply:
It is a pseudocode yes, the original looks like this:
![image](https://user-images.githubusercontent.com/94193030/142905628-f86ac8bf-e844-431b-ae13-7c1e233554b8.png)

Reply:
I guess we can leave it without syntax highlighting then 🤷